Camera user settings server

marktrayer opened the pull request:
#### Testing
- update all-clusters zap file with new cluster
- build all-clusters and chip-tool
- onnetwork pair all-clusters
- reads of all available attributes
- execution of all MPTZ related commands 
- all via chip-tool
- verify (visually) response in chip-too, and logs in all-clusters

marktrayer committed:
Very drafty initial server implementation for AV Settings User Level
Management.
For unit testing only.

marktrayer committed:
Updates to Camera AV Settings based on basic testing on the plumbing for read an command calls.
Command logic still TBD.

marktrayer committed:
Add skeletal handlers for read on complex typed attributes

marktrayer committed:
Implementations of all command handlers for Mechanical PTZ.  Implementation of simple delegate. Integration in All Clusters.
Basic unit testing via chip-tool -> all-clusters.

marktrayer committed:
Improved delegate handling.  All Manual PTZ commands implemented.

marktrayer committed:
Merge branch 'master' into camera-user-settings-server

marktrayer committed:
Rerun zap regen.
Re-add AV Settings to all-clusters.
All following merge of master.

marktrayer committed:
Add handling for the presence or otherwise of optional attributes. Have the delegate support all attributes.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

restyled-commits committed:
Restyled by prettier-json

marktrayer committed:
Add missing includes (note, this builds fine using the SDK scripts!).

marktrayer committed:
Addressing various CI failures

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Fix erroneous cast (int rather than uint)

marktrayer committed:
Be consistent with type use for zoom handling.  Mixed signed and unsigned was adding complexity and indeterminate results.
All values are now signed, with range checking applied.

marktrayer committed:
Merge branch 'camera-user-settings-server' of https://github.com/marktrayer/connectedhomeip into camera-user-settings-server

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Merging inadvertently backed out type change. Redo.

marktrayer committed:
Add use of static_cast to get past Tizen and ASR build issues.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+{
+    if (pan.HasValue())
+    {
+        mMptzPosition.pan = pan;
```
Needs to mark dirty, no?

Also, why is setting an empty optional ignored?

Similar questions for other setters.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        });
+        if (it == mMptzPresetHelper.end())
+        {
+            currentPresetID = static_cast<uint8_t>(currentPresetID + increment);
```
Why is `currentPresetID` not named like the other members?  `mCurrentPresetID` or so.

marktrayer replied:
Renamed.

marktrayer committed:
Improve attribute setting to allow for range checking of the values provided by the app.
Add minimal delegate and server handling for digital move.

marktrayer committed:
Add licence, and fix other license issues.

marktrayer committed:
Code comment resolution.
Address naming issues with features.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer commented on examples/all-clusters-app/all-clusters-common/include/camera-av-settings-user-level-management-instance.h:
```diff
@@ -0,0 +1,46 @@
+#pragma once
```
That's been added. Sorry for the omission.

marktrayer commented on examples/all-clusters-app/all-clusters-common/src/camera-av-settings-user-level-management-stub.cpp:
```diff
@@ -0,0 +1,131 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
```
Fixed.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.h:
```diff
@@ -0,0 +1,264 @@
+public:
+    virtual ~MPTZPresetHelper() = default;
+    MPTZPresetHelper() {}
+    MPTZPresetHelper(uint8_t aPreset, chip::CharSpan aName, MPTZStructType aPosition)
```
Fixed.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+{
+    uint8_t presetToRemove = commandData.presetID;
+
+    // Is the provided ID known to us?
```
Added in recent commits. Please confirm.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+{
+
+    // Call the delegate
+    Status status = mDelegate->DPTZSetViewport();
```
Updated in recent commits, that was a placeholder method signature.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+                                                                 const Commands::DPTZRelativeMove::DecodableType & commandData)
+{
+    // Call the delegate
+    Status status = mDelegate->DPTZRelativeMove();
```
Updated in recent commits, that was a placeholder method signature.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        VerifyOrReturnError(HasFeature(Feature::kMechanicalPresets), CHIP_ERROR_INVALID_ARGUMENT,
+                            ChipLogError(Zcl,
+                                         "CameraAVSettingsUserLevelManagement[ep=%d]: Feature configuration error. If MaxPresets "
+                                         "is enabled, then MaxPresets feature is required",
```
Fixed.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        VerifyOrReturnError(HasFeature(Feature::kDigitalPTZ), CHIP_ERROR_INVALID_ARGUMENT,
+                            ChipLogError(Zcl,
+                                         "CameraAVSettingsUserLevelManagement[ep=%d]: Feature configuration error. If "
+                                         "DPTZRelativeMove is enabled, then DPTZ feature is required",
```
Made it consistent in recent commits.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        VerifyOrReturnError(HasFeature(Feature::kMechanicalPresets), CHIP_ERROR_INVALID_ARGUMENT,
+                            ChipLogError(Zcl,
+                                         "CameraAVSettingsUserLevelManagement[ep=%d]: Feature configuration error. If MPTZPresets "
+                                         "is enabled, then MaxPresets feature is required",
```
Fixed.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        VerifyOrReturnError(HasFeature(Feature::kMechanicalZoom), CHIP_ERROR_INVALID_ARGUMENT,
+                            ChipLogError(Zcl,
+                                         "CameraAVSettingsUserLevelManagement[ep=%d]: Feature configuration error. If ZoomMax is "
+                                         "enabled, then Mechanical Zoom feature is required",
```
Fixed.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        VerifyOrReturnError(HasFeature(Feature::kMechanicalTilt), CHIP_ERROR_INVALID_ARGUMENT,
+                            ChipLogError(Zcl,
+                                         "CameraAVSettingsUserLevelManagement[ep=%d]: Feature configuration error. If TiltMin is "
+                                         "enabled, then Mechanical Tilt feature is required",
```
Fixed.

marktrayer committed:
Handle marking changed attributes as dirty

marktrayer committed:
Improve MPTZ Presets handling.  Add constraint check.  Allow for over-write of existing preset.

marktrayer committed:
Additional comment resolution.
Handle device being available to process an MPTZ change in the correct, spec'd location.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.h:
```diff
@@ -0,0 +1,264 @@
+     * Allows for a delegate or application to set the pan value given physical changes on the device itself, possibly due to direct
+     * user changes
+     */
+    void setPan(Optional<int16_t>);
```
Everything should now be FooBar casing.

marktrayer commented on src/app/zap-templates/zcl/data-model/chip/camera-av-settings-user-level-management-cluster.xml:
```diff
@@ -31,41 +31,18 @@ Git: 0.7-summer-2025-1-ge4654a5ef
-    <item fieldId="2" name="Zoom" type="int8u" optional="true" default="1" min="1" max="100"/>
+    <item fieldId="0" name="Pan" type="int16s" optional="true" default="0"/>
+    <item fieldId="1" name="Tilt" type="int16s" optional="true" default="0"/>
+    <item fieldId="2" name="Zoom" type="int8s" optional="true" default="1" min="1"/>
```
Reflects agreed spec change that's in Aron's omnibus PR: 
https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11333

marktrayer commented on src/app/zap-templates/zcl/data-model/chip/camera-av-settings-user-level-management-cluster.xml:
```diff
@@ -90,13 +67,13 @@ Git: 0.7-summer-2025-1-ge4654a5ef
       </mandatoryConform>
     </attribute>
-    <attribute code="0x0004" side="server" define="DPTZ_RELATIVE_MOVE" type="int8u" optional="true" min="2" max="100" default="100">
+    <attribute code="0x0004" side="server" define="ZOOM_MAX" type="int8s" optional="true" min="2" max="100" default="100">
```
Reflects agreed spec change that's in Aron's omnibus PR:
https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11333

marktrayer commented on src/app/zap-templates/zcl/data-model/chip/camera-av-settings-user-level-management-cluster.xml:
```diff
@@ -130,7 +107,7 @@ Git: 0.7-summer-2025-1-ge4654a5ef
       <arg id="0" name="Pan" type="int16s" optional="true" min="-180" max="180"/>
       <arg id="1" name="Tilt" type="int16s" optional="true" min="-180" max="180"/>
-      <arg id="2" name="Zoom" type="int8u" optional="true" min="1" max="100"/>
+      <arg id="2" name="Zoom" type="int8s" optional="true" min="1" max="100"/>
```
Reflects agreed spec change that's in Aron's omnibus PR:
https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11333

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+    // Set default MPTZ
+    mMptzPosition.pan  = defaultPan;
+    mMptzPosition.tilt = defaultTilt;
+    mMptzPosition.zoom = defaultZoom;
```
Reworked this area.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+            // Get the details to encode from the preset helper
+            //
+            MPTZPresetStructType presetStruct;
+            std::string aName = mptzPresets.GetName();
+            uint8_t aPreset   = mptzPresets.GetPresetID();
```
Renamed

bzbarsky-apple replied:
I don't see these being renamed.....

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+    Optional<int16_t> tilt = commandData.tilt;
+    Optional<int8_t> zoom  = commandData.zoom;
+
+    // Check with the delegate that we're in a position to change any of the PTZ values
```
Moved the check that results in Busy on a failure to align with the spec logic flow.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+    int16_t newTilt             = mMptzPosition.tilt.Value();
+    int8_t newZoom              = mMptzPosition.zoom.Value();
+
+    // Check with the delegate that we're in a position to change any of the PTZ values
```
Moved the check that results in Busy on a failure to align with the spec logic flow.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+    // Set the local values of pan, tilt, and zoom
+    setPan(Optional(static_cast<int16_t>(newPan)));
+    setTilt(Optional(static_cast<int16_t>(newTilt)));
+    setZoom(Optional(static_cast<int8_t>(newZoom)));
```
They've been removed.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+    // This is effectively a manipulation of the current PTZ settings, ensure that the device is in a state wherein a PTZ change is
+    // possible
+    //
+    if (!mDelegate->CanChangeMPTZ())
```
Moved this check, and added constraint checking.

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+
+    // Make sure that the vector will not exceed the max size
+    //
+    if (mMptzPresetHelper.size() == mMaxPresets)
```
Fixed the spec (in omnibus PR) and the ability to update an existing preset

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        }
+        int16_t panDeltaValue = panDelta.Value();
+        newPan                = static_cast<int16_t>(newPan + panDeltaValue);
+        if (newPan > mPanMax)
```
Added constraint checks to received delta values

marktrayer commented on src/app/zap-templates/zcl/data-model/chip/camera-av-settings-user-level-management-cluster.xml:
```diff
@@ -31,41 +31,18 @@ Git: 0.7-summer-2025-1-ge4654a5ef
   </struct>
 
-  <cluster apiMaturity="provisional">
+  <cluster>
```
Fixed

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,845 @@
+        return;
+    }
+
+    // Validate the received command fields
```
Should have now been added

marktrayer committed:
Comment resolution
- remove persistent storage code
- add constraint checks
- replace ConstraintError with NotFound where applicable

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.h:
```diff
@@ -236,15 +287,16 @@ class Delegate
-    virtual Protocols::InteractionModel::Status MPTZRemovePreset()                      = 0;
-    virtual Protocols::InteractionModel::Status DPTZSetViewport()                       = 0;
-    virtual Protocols::InteractionModel::Status DPTZRelativeMove()                      = 0;
+    virtual Protocols::InteractionModel::Status MPTZRelativeMove(Optional<int16_t> aPan, Optional<int16_t> aTilt,
```
Again: why the `using` for Status if we plan to just fully qualify it?  See also https://github.com/project-chip/connectedhomeip/pull/38170#discussion_r2021749199

marktrayer committed:
Merge branch 'master' into camera-user-settings-server

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Merge branch 'camera-user-settings-server' of https://github.com/marktrayer/connectedhomeip into camera-user-settings-server

marktrayer committed:
Errata corrections

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Compile error fix

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Revert Zoom to be a uint

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Address missing int -> uint handling

marktrayer committed:
Fix script change format

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.h:
```diff
@@ -0,0 +1,305 @@
+constexpr uint8_t kMaxZoomValue = 100;
+
+// Spec defined defaulta for Pan, Tilt, and Zoom
+constexpr int16_t defaultPan  = 0;
+constexpr int16_t defaultTilt = 0;
+constexpr uint8_t defaultZoom = 1;
```
```suggestion
constexpr int16_t kDefaultPan  = 0;
constexpr int16_t kDefaultTilt = 0;
constexpr uint8_t kDefaultZoom = 1;
```

marktrayer replied:
Fixed

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-settings-user-level-management-cluster.xml:
```diff
@@ -31,41 +31,18 @@ Git: 0.7-summer-2025-1-ge4654a5ef
   </struct>
 
-  <cluster apiMaturity="provisional">
+  <cluster apiMaturity="provisional">>
```
```suggestion
  <cluster apiMaturity="provisional">
```

marktrayer replied:
Fixed

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    }
+
+    mTiltMin = aTiltMin;
+    MarkDirty(Attributes::TiltMin::Id);
```
Should not MarkDirty if the value did not actually change..... That applies to all the setters.

marktrayer replied:
Fixed in this, and all the other set methods

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+//
+void CameraAvSettingsUserLevelMgmtServer::SetPan(Optional<int16_t> aPan)
+{
+    if (aPan.HasValue())
```
So what if the value is missing?  Shouldn't we still set it in mMptzPosition?  I'm not sure why this takes an Optional at all, but then ignores missing values...

This applies to a number of attributes.

Maybe this needs some documentation in the header explaining the behavior and why that behavior is the way it is?  It sounds like the goal is to have the member never have a missing value?

marktrayer replied:
Added a comment above these methods.  The value can be missing, it presence depends on feature flags.  An attempt to write a missing value can happen,e.g. an MPTZSetPosition with only one of the three being set.  This is basically ignored.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    do
+    {
+        auto it = std::find_if(mMptzPresetHelpers.begin(), mMptzPresetHelpers.end(), [=](const MPTZPresetHelper & mptzph) {
+            return mptzph.GetPresetID() == mCurrentPresetID + increment;
```
So I am not sure I follow this logic.  Say mCurrentPresetID == 3, mMaxPresets == 5, the presets with 2, 3, 4, 5 are in use.  Can that situation happen?

In that case, this code will try check for a preset with id 4, then id 5, then id 6... will set mCurrentPresetID to 6 and return.  Is that the expected behavior?  Some documentation on mCurrentPresetID that explains the state management might be nice.

marktrayer replied:
Hmm, thinking about this the code isn't handling wrap around to zero when we hit an ID that is mMaxPresets -1.  This will need rework.

marktrayer replied:
Reworked the code and tested via chip-tool

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        if (!HasFeature(Feature::kMechanicalPan))
+        {
+            ChipLogError(Zcl, "Mechanical Pan not supported although pan value provided");
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
```
```suggestion
            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand);
```
is what the spec says, no?

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        if (!HasFeature(Feature::kMechanicalTilt))
+        {
+            ChipLogError(Zcl, "Mechanical Tilt not supported although tilt value provided");
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
```
```suggestion
            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand);
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        if (!HasFeature(Feature::kMechanicalZoom))
+        {
+            ChipLogError(Zcl, "Mechanical Zoom not supported although zoom value provided");
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
```
```suggestion
            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand);
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    SetTilt(tilt);
+    SetZoom(zoom);
+
+    MarkDirty(Attributes::MPTZPosition::Id);
```
SetPan/Zoom/Tilt do that already as needed. Why do we need it here?

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        if (!HasFeature(Feature::kMechanicalPan))
+        {
+            ChipLogError(Zcl, "Mechanical Pan not supported although panDelta value provided");
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
```
```suggestion
            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand);
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+            return;
+        }
+        int16_t panDeltaValue = panDelta.Value();
+        if (panDeltaValue > (mPanMax - mPanMin) || panDeltaValue < -(mPanMax - mPanMin))
```
So this is assuming that `mPanMax - mPanMin` won't overflow?  I guess as long as all the systems this runs on have 4-byte `int` that will be true....  Might be worth static_asserting that the type of that difference is large enough, though (e.g. by storing in variable with `auto` type, then static_asserting something about the sizeof of that variable, then comparing panDelta to that variable).

marktrayer replied:
These are all int16_t, and the max possible value of mPanMax - mPanMin is 180 - -180 = 360.
So should be safe, unless I'm missing something?

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        }
+
+        newPan = static_cast<int16_t>(newPan + panDeltaValue);
+        if (newPan > mPanMax)
```
OK, so this is safe because the pan values are actually constrained so that panMax - panMin <= 360, right?  That might be worth a comment explaining why overflow can't happen here: `newPan` is between -180 and 180, and the thing being added is -360 to 360, so the sum can't overflow `int16_t`.

marktrayer replied:
Added comment.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        if (!HasFeature(Feature::kMechanicalTilt))
+        {
+            ChipLogError(Zcl, "Mechanical Tilt not supported although tiltDelta value provided");
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
```
```suggestion
            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand);
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+            return;
+        }
+        int16_t tiltDeltaValue = tiltDelta.Value();
+        if (tiltDeltaValue > (mTiltMax - mTiltMin) || tiltDeltaValue < -(mTiltMax - mTiltMin))
```
See comments for pan above.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        if (!HasFeature(Feature::kMechanicalZoom))
+        {
+            ChipLogError(Zcl, "Mechanical Zoom not supported although zoomDelta value provided");
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
```
```suggestion
            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand);
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    Optional<int8_t> zoomDelta  = commandData.zoomDelta;
+    int16_t newPan              = mMptzPosition.pan.Value();
+    int16_t newTilt             = mMptzPosition.tilt.Value();
+    int8_t newZoom              = static_cast<int8_t>(mMptzPosition.zoom.Value());
```
So this is safe because ZoomMax < 127, right?  Worth documenting.

marktrayer replied:
added a comment

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+            return;
+        }
+
+        newZoom = static_cast<int8_t>(newZoom + zoomDeltaValue);
```
OK, so here our spec constraints are:

* 2 <= ZoomMax <= 100.
* So -99 <= zoomDeltaValue <= 99.
* newZoom <= 100 before the addition.

That means this addition can in fact overflow (because 199 doesn't fit in `int8_t`), leading to incorrect behavior.  We clearly need some tests here.

We need to either make the result of the addition a type large enough to hold any necessary values, or rewrite this logic in some way.  For example:

1. Make newZoom a `uint8_t`.
2. If zoomDeltaValue < 0, check for `1 + static_cast<uint8_t>(-zoomDeltaValue) < newZoom` (before modifying the value of `newZoom`, else just set `newZoom` to `1`.
3. etc.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+
+    // Call the delegate to simply set the newly calculated MPTZ values based on the deltas received
+    //
+    status = mDelegate->MPTZRelativeMove(Optional(static_cast<int16_t>(newPan)), Optional(static_cast<int16_t>(newTilt)),
```
Why do you need those casts?  newPan and newTilt are already those types...  Just `MakeOptional(newPan)` etc.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    }
+
+    // Set the local values of pan, tilt, and zoom
+    SetPan(Optional<int16_t>(newPan));
+    SetTilt(Optional<int16_t>(newTilt));
```
Again, MakeOptional(newPan) etc.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    SetTilt(Optional<int16_t>(newTilt));
+    SetZoom(Optional<uint8_t>(static_cast<uint8_t>(newZoom)));
+
+    MarkDirty(Attributes::MPTZPosition::Id);
```
Not needed, since the setters handle it?

marktrayer replied:
Removed

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+
+    // Verify the provided presetID is within spec limits
+    //
+    if (preset > mMaxPresets - 1)
```
Doesn't match the spec in https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11333/files, so please just make sure that difference is tracked and this is fixed once the spec updates.

marktrayer replied:
Fixed in this PR per guidance from Aron.  Same for other comments on presetID checks

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    if (mMptzPresetHelpers.empty())
+    {
+        ChipLogError(Zcl, "No stored presets, MoveToPreset not possible");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
```
Again, https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11333/files is changing this; please track the difference.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    if (it == mMptzPresetHelpers.end())
+    {
+        ChipLogError(Zcl, "No matching presets, MoveToPreset not possible");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
```
And here.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    }
+
+    mMptzPosition = presetValues;
+    MarkDirty(Attributes::MPTZPosition::Id);
```
Should check that something actually changed before marking dirty?

marktrayer replied:
Reworked this

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    //
+    if (preset.HasValue())
+    {
+        if (preset.Value() > mMaxPresets - 1)
```
Spec differs after the omnibus PR, please track.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+
+    // Verify the provided presetID is within spec limits
+    //
+    if (presetToRemove > mMaxPresets - 1)
```
Might need changing depending on what happens with spec; please track.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+            return;
+        }
+        AddMoveCapableVideoStreamID(videoStreamID);
+        MarkDirty(Attributes::DPTZRelativeMove::Id);
```
AddMoveCapableVideoStreamID did that already, right?

marktrayer replied:
It did ... deleting!

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        int8_t zoomDeltaValue = zoomDelta.Value();
+        if (zoomDeltaValue < -100 || zoomDeltaValue > 100)
+        {
+            ChipLogError(Zcl, "Provided Digital Zoom Delta is out of range");
```
Perhaps log the value?  And maybe the endpoint....

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+            return;
+        }
+        AddMoveCapableVideoStreamID(videoStreamID);
+        MarkDirty(Attributes::DPTZRelativeMove::Id);
```
Done by AddMoveCapableVideoStreamID already.

marktrayer replied:
True ... deleted.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+        //
+        if (!mDelegate->IsValidVideoStreamID(videoStreamID))
+        {
+            ChipLogError(Zcl, "Unknown Video Stream ID provided.");
```
Log id and endpoint?

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1079 @@
+    // Call the delegate
+    Status status = mDelegate->DPTZRelativeMove(videoStreamID, deltaX, deltaY, zoomDelta);
+
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::Success);
```
This seems like a complicated way to write:
```suggestion
    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
```

marktrayer replied:
Fixed

marktrayer committed:
Add checks for feature flags on setting attributes
Improve preset ID setting and calculation
Better support the fact that pan, tilt, and zoom do not have to be all present (one of does)

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Change save to a preset to cover marking dirty correctly
Rework casts to get past compile issues

marktrayer committed:
Make all logs (errors and details) consistent in formatting

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Update comment headers and other documentation

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Align status responses with spec omnibus PR

yufengwangca commented on examples/all-clusters-app/ameba/chip_main.cmake:
```diff
@@ -157,7 +157,8 @@ list(
     ${chip_dir}/examples/all-clusters-app/all-clusters-common/src/air-quality-instance.cpp
-	${chip_dir}/examples/all-clusters-app/all-clusters-common/src/chime-instance.cpp
+    ${chip_dir}/examples/all-clusters-app/all-clusters-common/src/camera-av-settings-user-level-management-stub.cpp
+	  ${chip_dir}/examples/all-clusters-app/all-clusters-common/src/chime-instance.cpp
```
remove extra spaces

marktrayer replied:
Fixed in latest commit.

marktrayer committed:
Remove erroneous spaces

pidarped commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.h:
```diff
@@ -0,0 +1,350 @@
+private:
+    Delegate * mDelegate;
+    EndpointId mEndpointId;
+    BitMask<Feature> mFeatures;
```
Should be BitFlags given the argument passed to ctor is BitFlags.

pidarped commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.h:
```diff
@@ -0,0 +1,350 @@
+    Delegate * mDelegate;
+    EndpointId mEndpointId;
+    BitMask<Feature> mFeatures;
+    BitMask<OptionalAttributes> mOptionalAttrs;
```
BitFlags as well.

marktrayer committed:
Addressing code comments

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1219 @@
+/**
+ * Methods handling known video stream IDs, the addition and removal thereof.
+ */
+void CameraAvSettingsUserLevelMgmtServer::AddMoveCapableVideoStreamID(uint16_t aVideoStreamID)
```
Removed the check

marktrayer commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1219 @@
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+            return;
+        }
+        AddMoveCapableVideoStreamID(videoStreamID);
```
Changed the logic so that a relative move is only possible via a known video stream id, provided via SetViewport

marktrayer committed:
Reflect that MaxPresets is now F quality
Fix logic in DPTZRelativeMove

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -661,103 +766,129 @@ void CameraAvSettingsUserLevelMgmtServer::HandleMPTZRelativeMove(HandlerContext
 
-        newZoom = static_cast<int8_t>(newZoom + zoomDeltaValue);
-        if (newZoom > mZoomMax)
+        // If we're here, then we'll also have an existing Pan value in MPTZPosition. The zoom values are constrained such that
```
```suggestion
        // If we're here, then we'll also have an existing Zoom value in MPTZPosition. The zoom values are constrained such that
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -661,103 +766,129 @@ void CameraAvSettingsUserLevelMgmtServer::HandleMPTZRelativeMove(HandlerContext
+        // If we're here, then we'll also have an existing Pan value in MPTZPosition. The zoom values are constrained such that
+        // we won't overflow newZoomValue
+        //
+        int16_t newZoomValue = static_cast<int16_t>(mMptzPosition.zoom.Value() + zoomDeltaValue);
```
So this is relying on some pretty subtle promotion mechanics (e.g. that the unsigned value gets promoted to a signed one automatically, because the unsigned type is sufficiently narrow, etc).  Might clearer as:
```suggestion
        int newZoomValue = static_cast<int>(mMptzPosition.zoom.Value()) + zoomDeltaValue;
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -800,19 +929,18 @@ void CameraAvSettingsUserLevelMgmtServer::HandleMPTZMoveToPreset(HandlerContext
     // Verify the provided presetID is within spec limits
     //
-    if (preset > mMaxPresets - 1)
+    if (preset > mMaxPresets)
```
OK, but also need to ensure that preset >= 1, no?

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -869,9 +1001,10 @@ void CameraAvSettingsUserLevelMgmtServer::HandleMPTZSavePreset(HandlerContext &
     if (preset.HasValue())
     {
-        if (preset.Value() > mMaxPresets - 1)
+        if (preset.Value() > mMaxPresets)
```
Need to ensure >= 1, if we are following the not-yet-merged spec PR.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -823,8 +951,10 @@ void CameraAvSettingsUserLevelMgmtServer::HandleMPTZMoveToPreset(HandlerContext
-        ChipLogError(Zcl, "No matching presets, MoveToPreset not possible");
-        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        ChipLogError(Zcl,
+                     "CameraAVSettingsUserLevelMgmt[ep=%d]: No matching presets, MoveToPreset not possible for provide preset: %d",
```
```suggestion
                     "CameraAVSettingsUserLevelMgmt[ep=%d]: No matching presets, MoveToPreset not possible for provided preset: %d",
```

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -945,9 +1080,10 @@ void CameraAvSettingsUserLevelMgmtServer::HandleMPTZRemovePreset(HandlerContext
     // Verify the provided presetID is within spec limits
     //
-    if (presetToRemove > mMaxPresets - 1)
+    if (presetToRemove > mMaxPresets)
```
Need to ensure >= 1?

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1219 @@
+    // Call the delegate
+    Status status = mDelegate->DPTZRelativeMove(videoStreamID, deltaX, deltaY, zoomDelta);
+
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
```
```suggestion
```

marktrayer committed:
Handle comments
Constraint checks
Fix handling of PresetID, and logic around when a new preset is added (to ensure constraint check when required)

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Fix up casts for zoom delta handling

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1196 @@
+    Optional<uint8_t> preset  = commandData.presetID;
+    chip::CharSpan presetName = commandData.name;
+    uint8_t presetToUse       = mCurrentPresetID;
+    bool newPresetValue       = true;
```
Two things:

1.  Not sure why this is declared all the way up here, not where it used to be declared below where it's first computed for real.
2. The logic setting this value seems pretty complicated.  I wonder whether it would be better to reverse the sense of this boolean, call it `updatingExistingPreset` or something, and then `bool updatingExistingPreset = preset.HasValue() && it != mMptzPresetHelpers.end();`, which makes it pretty clear what's going on: a preset value was provided, and it exists.

bzbarsky-apple commented on src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp:
```diff
@@ -0,0 +1,1196 @@
+    }
+
+    // Update the current preset ID to the next available only if we actually used the current value.  A user provided preset
+    // could any value between 1 and MaxPresets
```
```suggestion
    // could have any value between 1 and MaxPresets
```

marktrayer replied:
Fixed

marktrayer committed:
Address code comments.
Simplify preset save handling.
Improve logging.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

marktrayer committed:
Undo accidental removal of attributes from zap
